### ISSUE TITLE
Consider making built-in block and inline macro handling replaceable

### ISSUE THREAD

**Issue by djencks** [asciidoctor/asciidoctor#3562]
There are several builtin inline-macro-like functionalities such as 

* stem:
* xref:
* image:

and block like functionalities such as 

* [stem]

that currently can't be directly replaced.  Stem can be dealt with using a fairly complicated treeProcessor approach; Antora uses some Opal magic to replace xref: and inline images.

Making these sorts of built in functionality replaceable would make some kinds of extension of asciidoctor considerably simpler.

Since these all look like blocks or inline  macros, which fit into the existing extension mechanism, one obvious possibility would be to implement the built-in functionality as extensions, and have that as the default; registering an explicit extension for say "stem" would replace the default.

I'm not sure if the syntax of e.g. stem: is exactly that of an inline macro (I haven't been able to write an inline macro that has nothing between : and [, such as stem:[x^2]), but I imagine massaging the input if necessary would not be difficult.

**Comment by djencks** [comment]
Roughly this capability was requested in https://github.com/asciidoctor/asciidoctor/issues/3474

**Comment by mojavelinux** [comment]
What you're asking is a very broad issue and could have other consequences we need to consider. It might be better to request a stem adapter. You don't really want to parse the macro. What you want is to be able to manipulate the content that it captures, particularly stem. asciidoctor-mathematical also has this need.

**Comment by djencks** [comment]
Let's concentrate first on stem.
I don't understand if `stem:` syntax falls into what's allowed for an inline macro, but I think that the processing can easily go through the inline macro interface, and [stem] block processing can go through a block interface.  Perhaps I'm suggesting that a stem adapter can use the existing extension interfaces?

I have some evidence for thinking the content can be massaged to go to an inline macro, I have a tree processor that does that, but cannot deal with substitutions. I'm thinking that in place of the existing stem: code, the data picked up could be transformed slightly to be input to an inline macro, and sent to one.  The default stem inline macro implementation would produce the current behavior.

**Comment by mojavelinux** [comment]
The reason the stem inline macro cannot easily be replaced is because it has special semantics. It needs to passthrough the content, so it gets processed as a pass macro (and pass macros cannot be nested). So for all intents and purposes, it's not a normal inline macro.

**Comment by mojavelinux** [comment]
There's a FIXME in the extensions code that reads: "for inline passthrough, we need to have some way to specify the text as a passthrough", so I have thought about the need for this kind of a custom inline passthrough macro before.

**Comment by djencks** [comment]
Yes.... so without allowing "passthrough inline macros" in general, stem: processing can't just be an inline macro.  However, after that special subs processing, and anything else special that it needs, I think the call to "process this stuff!" can go to something implemented as an inline macro.  If "passthrough inline macros" are ever implemented, then perhaps stem processing can be one of them with nothing else special about them.

Too bad the comments in ruby don't get into the opalescent javascript code :-)
